File: Problems/0002_Add_Two_Number/Project_CS/Add_Two_Number.cs
using System;

//Definition for singly-linked list.
public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int x) { val = x; }
}

public class Solution
{
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2)
    {
        ListNode dummyHead = new ListNode(0);
        ListNode p = l1, q = l2, curr = dummyHead;
        int carry = 0;

        while (p != null || q != null)
        {
            int x = (p != null) ? p.val : 0;
            int y = (q != null) ? q.val : 0;
            int sum = carry + x + y;
            carry = sum / 10;
            curr.next = new ListNode(sum % 10);
            curr = curr.next;
            if (p != null)
                p = p.next;
            if (q != null)
                q = q.next;
        }

        if (carry > 0)
        {
            curr.next = new ListNode(carry);
        }

        return dummyHead.next;        
    }

    public int[] str_to_int_array(string s)
    {
        string[] flds = s.Split(',');
        int[] nums = new int[flds.Length];

        if (flds.Length <= 0)
            return nums;

        for (int i = 0; i < nums.Length; ++i)
        {
            nums[i] = int.Parse(flds[i]);
        }

        return nums;
    }

    public string output_int_array(int[] nums)
    {
        if (nums.Length <= 0)
            return "";

        string resultStr = "[" +  nums[0].ToString();
 
        for (int i = 1; i < nums.Length; ++i)
        {
            resultStr += "," + nums[i].ToString();
        }

        return resultStr + "]";
    }

    public ListNode set_nodes(int[] nums, int index)
    {
        if (index >= nums.Length)
            return null;
        
        ListNode node = new ListNode(nums[index]);
        node.next = set_nodes(nums, index + 1);

        return node;
    }

    public string output_nodes(ListNode ll)
    {
        string retStr = ll.val.ToString(); 

        if (ll.next != null)
            retStr += " -> " + output_nodes(ll.next);

        return (retStr);
    }

    public void Main(string args)
    {
        string arg_str = args.Replace("[[","").Replace("]]","").Trim();
        string[] flds = arg_str.Split(new string[] {"],["}, StringSplitOptions.None);
        int[] num1 = str_to_int_array(flds[0]);
        int[] num2 = str_to_int_array(flds[1]);
        Console.WriteLine("num1 = " + output_int_array(num1));
        Console.WriteLine("num2 = " + output_int_array(num2));

        ListNode l1 = set_nodes(num1, 0);
        ListNode l2 = set_nodes(num2, 0);
        Console.WriteLine("l1 = " + output_nodes(l1));
        Console.WriteLine("l2 = " + output_nodes(l2));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        ListNode result = AddTwoNumbers(l1, l2);
        Console.WriteLine("result = " + output_nodes(result));
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
    }
}

File: Problems/0001_0099/0002_Add_Two_Number/Project_C/Add_Two_Number.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>
#include <time.h>

#include "../../../mylib_C/mylib.h"

/* Function prototype declaration */
int* twoSum(int* nums, int numsSize, int target);
int loop_main(char* arg);

/* Definition for singly-linked list. */
struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2)
{
    struct ListNode* dummyHead = (struct ListNode*)malloc(sizeof(struct ListNode));
    dummyHead->val = 0;
    dummyHead->next = NULL;
    struct ListNode* p = l1;
    struct ListNode* q = l2;
    struct ListNode* curr = dummyHead;
    int carry = 0;

    while (p != NULL || q != NULL)
    {
        int x = (p != NULL) ? p->val : 0;
        int y = (q != NULL) ? q->val : 0;
        int sum = carry + x + y;

        carry = sum / 10;
        curr->next = (struct ListNode*)malloc(sizeof(struct ListNode));
        curr->next->val = sum % 10;
        curr->next->next = NULL;
        curr = curr->next;

        if (p != NULL)
            p = p->next;
        if (q != NULL)
            q = q->next;
    }

    if (carry > 0)
    {
        curr->next = (struct ListNode*)malloc(sizeof(struct ListNode));
        curr->next->val = carry;
        curr->next->next = NULL;
    }

    return dummyHead->next;
}

struct ListNode* set_nodes(int nums[], int nums_Size, int index)
{
    if (index >= nums_Size)
        return NULL;
    
    struct ListNode* node = (struct ListNode*)malloc(sizeof(struct ListNode));

    node->val = nums[index];
    node->next = set_nodes(nums, nums_Size, index + 1);

    return node;
}


char* output_nodes(struct ListNode* ll)
{
    #define resultStr_MAX_SIZE  1024

    char* resultStr = (char *)malloc(sizeof(char)*resultStr_MAX_SIZE);
    sprintf(resultStr, "%d", ll->val); 

    if (ll->next != NULL) {
        strcat(resultStr, " -> ");
        strcat(resultStr, output_nodes(ll->next));
    }

    return resultStr;
}

int ListNode_free(struct ListNode* ll)
{
    if (ll == NULL)
        return 0;
    
    int result = 0;

    if (ll->next != NULL) {
        result += ListNode_free(ll->next);
    }

    free(ll);
    result++;

    return result;
}

int loop_main(char* arg)
{
    #define nums_MAX_SIZE   256

    int argv_length = strlen(arg);
    char* flds[2];
    char* str_nums1[nums_MAX_SIZE];
    char* str_nums2[nums_MAX_SIZE];
    int nums1[nums_MAX_SIZE];
    int nums2[nums_MAX_SIZE];

    replace(arg, "[[", "");
    replace(arg, "]]", "");
    replace(arg, "\n", "");
    int flds_length = split(arg, "],[", flds, sizeof(flds)/sizeof(flds[0]));
    int str_nums1_length = split(flds[0], ",", str_nums1, sizeof(str_nums1)/sizeof(str_nums1[0]));
    int str_nums2_length = split(flds[1], ",", str_nums2, sizeof(str_nums2)/sizeof(str_nums2[0]));

    // int nums1[] size check.
    if (sizeof(nums1)/sizeof(nums1[0]) < str_nums1_length)
        err_exit("nums1[] size error.");

    // int nums2[] size check.
    if (sizeof(nums2)/sizeof(nums2[0]) < str_nums2_length)
        err_exit("nums2[] size error.");

    int nums1_length = str_to_int_array(str_nums1, nums1, str_nums1_length);
    int nums2_length = str_to_int_array(str_nums2, nums2, str_nums2_length);
    struct ListNode* l1 = set_nodes(nums1, nums1_length, 0);
    struct ListNode* l2 = set_nodes(nums2, nums2_length, 0);
    printf("l1 = %s\n", output_nodes(l1));
    printf("l2 = %s\n", output_nodes(l2));

    clock_t time_start = clock();
    struct ListNode* result = addTwoNumbers(l1, l2);
    clock_t time_end = clock();

    // result print.
    printf("result = %s\n", output_nodes(result));
    printf("Execute time ... %.0f ms\n\n", 1000*(double)(time_end - time_start)/CLOCKS_PER_SEC);

    // struct ListNode *l1, *l2, *result clear.
    int ret_result = ListNode_free(result);
    int ret_l1 = ListNode_free(l2);
    int ret_l2 = ListNode_free(l1);

    // char* str_nums1[], str_nums2[] free().
    p_char_array_free(str_nums2, str_nums2_length);
    p_char_array_free(str_nums1, str_nums1_length);

    // char* flds[] free().
    p_char_array_free(flds, flds_length);

    return 0;
}

int main(int argc, char* argv[])
{
#define fgets_MAX   65536
    FILE *fp;
    char line[fgets_MAX];

    if (argc < 2) {
        printf("Usage %s <testdatafile>\n", argv[0]);
        return -1;
    }

    // File Open
    fp = fopen(argv[1], "r");

    if (fp == NULL) {
        printf("%s Open Failed...\n", argv[1]);
        return -1;
    }

    while((fgets(line, fgets_MAX - 1, fp)) != NULL) {
        trim(line);
        if (*line == '\0')
            continue;
        printf("arg = %s\n", line);
        loop_main(line);
    }

    fclose(fp);
    return 0;
}
